### PR TITLE
Allow BaseExperiment to accept experiment configuration directly

### LINKED ISSUE #10: Proposal: allow BaseExperiment to accept experiment configuration directly
Currently the only way to pass the experiment configuration to BaseExperiment is to use the `exp_config_file` argument in the constructor. The proposal is to add a new argument called `exp_config` of type dict. This argument, if not None, will contain the full experiment configuration as expected with `exp_config_file`. The difference is that with `exp_config` there is no need to read a JSON file from disk.

https://github.com/FFede0/RL4CC/blob/ac2c8e2891e166ca75eca6934d8833d7da039aac/experiments/base_experiment.py#L26-L35

What about the `env_config_file` and `ray_config_file` fields in the `exp_config` dictionary? I propose to add two more fields:

* `env_config`: a dictionary, if not None, must contain a full environment configuration as expected when reading the `env_config_file`;
* `ray_config`: a dictionary, if not None, must contain a complete ray configuration as expected when reading the `ray_config_file`;

If the user gives to RL4CC an experiment configuration with both `env_config_file` and `env_config`, it will throw an error. The same behaves with `ray_config_file` and `ray_config`.

If the user passes both the `exp_config_file` and `exp_config` arguments to BaseExperiment, the constructor will throw an error.

Examples of valid experiment configurations are:

<details>
<summary>exp_config given as dictionary, with env_config and ray_config given as files</summary>

```python
{
    "algorithm": "PPO",
    "env_config_file": "env_config.json",
    "ray_config_file": "ray_config.json",
    "stopping_criteria": {
        "max_iterations": 1
    }
}
```
</details>

<details>
<summary>exp_config given as dictionary, with env_config given as dictionary and ray_config given as files</summary>

```python
{
    "algorithm": "PPO",
    "env_config": {
        "env_name": "TrafficManagementEnv",
        "min_time": 0,
        "max_time": 100,
        "time_step": 1,
        "scenario": "scenario1"
     },
    "ray_config_file": "ray_config.json",
    "stopping_criteria": {
        "max_iterations": 1
    }
}
```
</details>

<details>
<summary>exp_config given as dictionary, with env_config and ray_config given as dictionaries</summary>

```python
{
    "algorithm": "PPO",
    "env_config": {
        "env_name": "TrafficManagementEnv",
        "min_time": 0,
        "max_time": 100,
        "time_step": 1,
        "scenario": "scenario1"
     },
    "ray_config_file": {
        "framework": "torch",
        "callbacks": {
            "callbacks_class": "RL4CC.callbacks.base_callbacks.BaseCallbacks"
        },
        "rollouts": {
            "num_rollout_workers": 1
        },
    },
    "stopping_criteria": {
        "max_iterations": 1
    }
}
```
</details>

<details>
<summary>exp_config given as JSON file, with env_config and ray_config given as dictionaries</summary>

```bash
$ cat exp_config.json
{
    "algorithm": "PPO",
    "env_config": {
        "env_name": "TrafficManagementEnv",
        "min_time": 0,
        "max_time": 100,
        "time_step": 1,
        "scenario": "scenario1"
     },
    "ray_config_file": {
        "framework": "torch",
        "callbacks": {
            "callbacks_class": "RL4CC.callbacks.base_callbacks.BaseCallbacks"
        },
        "rollouts": {
            "num_rollout_workers": 1
        },
    },
    "stopping_criteria": {
        "max_iterations": 1
    }
}
```
</details>

**Reason for the proposal**: I need to run a lot of experiments with small changes between each one. So it would be faster to dynamically generate the experiment configuration and pass it to RL4CC, rather than writing the configuration to a temporary file and passing the path to RL4CC.

**Impact and implementation**: minor changes in BaseExperiment, all backwards compatible.

What do you think? If accepted, I can try to implement the feature.

### PR DESCRIPTION
**Pull request by ema-pe**
This PR implements the feature discussed in #10.

To test this feature, you can download the attached sample project that calls TrainingExperiment with different combinations of exp_config, exp_config_file (and env_config and ray_config). You need to have RL4CC installed (and the `test-ep-10` branch) in the RL4CC directory of the example project. To run the tests, do `python main.py`.

Project: [test-ep-10.zip](https://github.com/user-attachments/files/15561196/test-ep-10.zip)

Closes #10.

~~Edit: I forgot to document the feature. I'll add a new commit in the next few hours.~~ Done.

### CODE REVIEW

**Comment by ema-pe** [comment]
**Question**: should I allow `tune_config` to be given directly instead of using `tune_config_file`?

https://github.com/FFede0/RL4CC/blob/a14bd81ac2c12aacbe5df26c4df8fea76d0ea6d3/experiments/tune.py#L36-L44

https://github.com/FFede0/RL4CC/blob/test/config_files/README.md#tuner-configuration-file

**Comment by FFede0** [comment]
> **Question**: should I allow `tune_config` to be given directly instead of using `tune_config_file`?
> 
> https://github.com/FFede0/RL4CC/blob/a14bd81ac2c12aacbe5df26c4df8fea76d0ea6d3/experiments/tune.py#L36-L44
> 
> https://github.com/FFede0/RL4CC/blob/test/config_files/README.md#tuner-configuration-file

To be coherent, I’d say yes

**Comment by ema-pe** [comment]
> > **Question**: should I allow `tune_config` to be given directly instead of using `tune_config_file`?
> > https://github.com/FFede0/RL4CC/blob/a14bd81ac2c12aacbe5df26c4df8fea76d0ea6d3/experiments/tune.py#L36-L44
> > 
> > https://github.com/FFede0/RL4CC/blob/test/config_files/README.md#tuner-configuration-file
> 
> To be coherent, I’d say yes

Done. I also updated the documentation to be coherent. I tested the code for TrainingExperiment, but not for TuningExperiment.

You can see the updated documentation here: https://github.com/FFede0/RL4CC/tree/test

File: README.md [hunk 1]
```diff
@@ -133,108 +136,123 @@ experiment, your `main.py` file should include:
...
-`exp_config.json` are simply copied from the user-specified input directory, 
-the `ray_config.json` file reported here includes also the default values 
-assigned to keys that were not included in the user-defined configuration file.
+The outputs produced during the training experiment are saved in a suitable
+sub-directory of the `logdir` specified in the [`exp_config`
+config](config_files/README.md#experiment-configuration) (or in `~/ray_results`
+if nothing is provided). These include:
+
+- `complete_config`: a directory containing the configuration (`exp_config`,
+  `envconfig` and `ray_config`) used to define the experiment, saved as JSON
```
Comment (FFede0) [inline README.md]: Typo: `env_config`

File: README.md [hunk 2]
```diff
@@ -133,108 +136,123 @@ experiment, your `main.py` file should include:
...
+  files.
+> [!NOTE]
+> Two important notes:
+>   - The configuration is saved as files even if you pass to RL4CC dictionaries
```
Comment (FFede0) [inline README.md]: JSON files are saved here, regardless the fact that the user passed the configuration(s) as file(s) or as `dict` object(s)

File: config_files/README.md [hunk 3]
```diff
@@ -89,16 +101,16 @@ the two approaches cannot be mixed.
 
 :warning::warning::warning: **Important note:** there are few elements that,
```
Comment (FFede0) [inline config_files/README.md]: I'm bothered by the coexistence of different formats. Please, write these as your "warnings"

File: experiments/base_experiment.py [hunk 4]
```diff
@@ -65,21 +75,47 @@ def validate_experiment_configuration(self):
...
+      # Load the env_config. The user can specify the env_config via the
+      # env_config_file parameter or directly via the env_config parameter.
+      if (not_defined("env_config_file", self.exp_config)
+          and not_defined("env_config", self.exp_config)):
+        raise KeyError("ERROR: provide 'env_config_file' or 'env_config' if no previous checkpoint is given")
+      if (defined("env_config_file", self.exp_config)
+          and defined("env_config", self.exp_config)):
+        raise KeyError("ERROR: 'env_config_file' or 'env_config' cannot be both set!")
+
+      if defined("env_config_file", self.exp_config):
```
Comment (FFede0) [inline experiments/base_experiment.py:99]: It makes sense to have defined...but I'd remove not_defined, then, or implement it as `return not defined(...)`

File: utilities/common.py [hunk 5]
```diff
@@ -25,6 +25,12 @@ def not_defined(param: str, params_dict: dict) -> bool:
   """
   return (param not in params_dict) or (params_dict[param] is None)
 
+def defined(param: str, params_dict: dict) -> bool:
```
Comment (FFede0) [inline utilities/common.py:28]: See comment to base_experiment.py

**Review by FFede0 — CHANGES_REQUESTED** [review]
Address comments to single files

**Review by FFede0 — APPROVED** [review]

**Pull request merged**